lib/Canonicalize/MakeModelCastPass.cpp

```diff
@@ -133,37 +133,35 @@ MMCP::serializeTypesForModelCast(FunctionMetadataCache &Cache,
       SerializeTypeFor(Ret->getOperandUse(0));
 
   } else if (auto *SI = dyn_cast<StoreInst>(I)) {
-    auto &PtrOperandPtrType = TypeMap.at(SI->getPointerOperand());
-    auto &ValOperandType = TypeMap.at(SI->getValueOperand());
-
-    const model::Architecture::Values &Arch = Model.Architecture();
-    QualifiedType ValOperandPtrType = ValOperandType.getPointerTo(Arch);
-    if (PtrOperandPtrType != ValOperandPtrType) {
-      bool IsPtrOperandOfAggregateType = false;
-      QualifiedType PtrOperandType;
-
-      if (PtrOperandPtrType.isPointer()) {
-        PtrOperandType = dropPointer(PtrOperandPtrType);
-        const auto *Unqual = PtrOperandType.UnqualifiedType().getConst();
-        IsPtrOperandOfAggregateType = isa<model::StructDefinition>(Unqual)
-                                      || isa<model::UnionDefinition>(Unqual);
-      }
+    auto &PtrOperandPtrType = *TypeMap.at(SI->getPointerOperand());
+    auto &ValOperandType = *TypeMap.at(SI->getValueOperand());
+
+    bool IsPointerToAnAggregate = false;
+    const model::UpcastableType *PointeeType = nullptr;
+    if (auto Pointer = PtrOperandPtrType.tryGetAsPointer()) {
+      PointeeType = &Pointer->PointeeType();
+      if (*Pointer->PointeeType() == ValOperandType)
+        return Result;
+
+      auto *Def = (*PointeeType)->tryGetAsDefinition();
+      IsPointerToAnAggregate = Def
+                               && (llvm::isa<model::StructDefinition>(Def)
+                                   || llvm::isa<model::UnionDefinition>(Def));
```

Comment:
1. We seem to talking about aggregates, but we're not handling arrays. cc/@pfez
2. Unrelated, we have `isScalar`, shall we have `isAggregate`?
   Note: we might have discussed in the past, but I'd expect, the following:
   * `isScalar(void)`: abort
   * `isAggregate(void)`: abort
   * Other then the above `isScalar(X) == not isAggregate(X)`

Reply:
1. No idea, maybe the fact that this is store instruction specific has something to do with it?
2. 👌

Reply:
1. I'm pretty sure today the whole `if (auto *SI = dyn_cast<StoreInst>(I))` block is now dead code. With some `git blame` magic I traced it back to the very first commit that introduced `MakeModelCast`, when from my `git log` we still had load and store instructions reaching the end of the pipeline. Nowadays we strip them out and convert them to `Copy`/`Assign` earlier than `MakeModelCast`. So that whole part is bogus.

Reply:
Should I try dropping it? Or just leave a todo?

Reply:
@pfez, ping

Reply:
Drop it. But do it in a separate commit.

Reply:
Done (both 1 and 2).